### ISSUE TITLE
rectangle-range-query for scipy.spatial.KDTree

### ISSUE THREAD

**Issue by tianfudhe** [scipy/scipy#7822]
As there're already circle-range-query, near-pair query, and NN-query supported in scipy.spatial.KDTree, why not add rectangle or even polygon range query in it? That would be great.

**Comment by rgommers** [comment]
@tianfudhe do you have a concrete use case and/or are interested in implementing this?

**Comment by tianfudhe** [comment]
@rgommers yes.
I'm analyzing some spatial data using Python and I need rectangle range query for simple visualization. 
the data is large enough to reject O(N) selection, so I seek for spatial index library, but find few stable libraries.
Python is a sharp tool for fast data analysis. it's a pity that so few tools support even such fundamental functionality.

**Comment by rgommers** [comment]
@rainwoodman @sturlamolden does this make sense to you to add, and should we therefore keep this issue open as an enhancement request?

**Comment by sturlamolden** [comment]
Scanning a kd-tree to do what @tianfudhe suggests is easy. I am not sure if it belongs in SciPy though, as it can be easily done with some lines of Cython or C++.

**Comment by sturlamolden** [comment]
Basically just start at the root node and use ```mins``` and ```maxes``` to check if the two child nodes are inside the rectangle. Then recurse down to the leaf nodes and check if the points in the leaves are inside the rectangle.

**Comment by rgommers** [comment]
> as it can be easily done with some lines of Cython or C++.

for a lot of users that can be quite a barrier. so it's more the question if there are enough use cases. I suspect so.

**Comment by sturlamolden** [comment]
Just leave the feature request open then.

**Comment by rainwoodman** [comment]
We are talking about extending the single tree algorithm range queries.  

The single tree rectangular / arbitrary shape query can be thought of a quick way of doing range integrals on a random sample from a density function.

Polygons are used to represent areas of sky that has been observed. http://casa.colorado.edu/~ajsh/mangle/

I don't think scipy has the data model to represent high dimensional regions. This might be the main barrier for adding polygon support (2-d is easy, but we have a `k-d tree`). high dimensional rectangular shapes are easier.

**Comment by rainwoodman** [comment]
I think shall first extend query_ball_point to accept vector `r` input (only an interface extension) Then one can use current cKDTree to prune for the rectangular queries in bulk before refining with rectangular cuts (which can also be vectorized).

**Comment by sturlamolden** [comment]
No, ```query_ball_point``` queries a ball around a point. It will be confusing to let it query rectangular regions.

**Comment by pyeguy** [comment]
what's the status on this? I have a pressing use case that would benefit ( i think ) from such a feature. 

my use case: 
i have m dimensional data (m~4) where each point has an associated error range (different error function for each dimension)

I want to find all points with overlapping error ranges and average their values.

**Comment by rainwoodman** [comment]
Would query_ball_point with p=1 work?

**Comment by rainwoodman** [comment]
@sturlamolden, query_ball_point already query a square region if p is set to 1. So allowing vector `r` would be a natural extension.

**Comment by pyeguy** [comment]
@rainwoodman  yes in my case they are distinctly not square/hypercubic.

I think @sturlamolden 's point of having the naming of `query_ball_point` refering to a hypersphere makes sense even if you can force it to be square with `p=inf`

Why not just add a `query_range` method? It would be great if accepted an array of ranges like `query_ball_point`

it could match the api for the python [rtree package](http://toblerity.org/rtree/tutorial.html#intersection)

**Comment by rainwoodman** [comment]
You must be mentioning the intersection function?

For vectored input, then we shall expect a (Nr, Nd, 2) array?
Perhaps receiving top = (..., Nd) and bottom = (..., Nd) would be better.

At the end of the day someone has to add the coding though -- do you want
to take a stab?
I may not have the bandwidth to focus on this recently.

As a separate note, query_ball_point and query_range shall probably be able
to return something that can be directly used with numpy.ufunc.reduceat.
I don't think we currently do that yet.

**Comment by pyeguy** [comment]
i took a stab at a python implementation. i'll see if I can wrap my head around the cKDTree internals enough to translate it but i'm a serious c/c++/cython noob so help or guidance on that front would be appreciated if the python implementation looks decent. 

also I've come to realize that I really want an R-Tree for my problem. any thoughts on expanding the set of data-structures in `scipy.spatial` ?

**Comment by rainwoodman** [comment]
I think you will need to add a function similar to this:

https://github.com/scipy/scipy/blob/master/scipy/spatial/ckdtree/src/ckdtree_methods.h#L190

searching for the function name shows you places where modifications shall be done:

https://github.com/scipy/scipy/search?q=query_ball_point&unscoped_q=query_ball_point

The is the ball point algorithm implementation:

https://github.com/scipy/scipy/blob/6f5ea33175671f634cef8a980656701d50becf12/scipy/spatial/ckdtree/src/query_ball_point.cxx

Notice that the C++ code combines recursion with a stack in tree traversal. This is tricky and I do not know why it was written like that. The good news is that I think you can reuse the RectRectDistanceTracker -- if the given rect intersects the node, I believe min_distance will be zero. We can worry about detection of the case where the query rect fully encloses the node rect later. I think to handle that you will need a new DistanceTracker class.

The code is written in this way with templates in order to support optimizations of p-norm. In the range query case, any p should work.

**Comment by JesseLivezey** [comment]
Is the centered-rectangle case equivalent to rescaling the features/spatial-coordinates by 1/(corresponding rectangle dimension) and then using query_ball_point(..., r=1, p=1)?

**Comment by davidkong0987** [comment]
Are there any simple workarounds for us to achieve this ouselves?

**Comment by sturlamolden** [comment]
> Are there any simple workarounds for us to achieve this ourselves?

The are at least two workarounds:

One is to simply write a Python, Cython or C++ function that does this. The kd-tree structure is fully exposed so just write a function that does the query.

Another is to scale the data. Scale your data so your hyperrectangle becomes a hypersquare. Then construct a new kd-tree and query ball point with p=1.

**Comment by davidcian** [comment]
I have a similar problem that would require a range query where the region would be a hyperrectangle and not a hypercube, I'm rather surprised that a pre-existing implementation is hard to find.

**Comment by sturlamolden** [comment]
> I have a similar problem that would require a range query where the region would be a hyperrectangle and not a hypercube, I'm rather surprised that a pre-existing implementation is hard to find.

The more general solution is to scale the data. This also allows for the cases where a hyperrectangle is not oriented in parallel with the axes, in which case one could also multiply the data with a rotation matrix.

**Comment by mdhaber** [comment]
@rainwoodman @sturlamolden @nickodell @MLopez-Ibanez given the recent momentum on KDTree, I wondered if I could put this enhancement request for rectangular regions on your radar. It has more thumbs up than any issue I can recall. Re: https://github.com/scipy/scipy/issues/7822#issuecomment-328345275, I agree with Ralf that it's fair game for SciPy. I'd be willing to review by test if someone is willing to write the code and someone else can take a look at it. @ev-br also took a look at this recently, and it is something that may be easy for those already familiar with the code, but would require quite a deep dive for the rest of us.

**Comment by sturlamolden** [comment]
> I'm analyzing some spatial data using Python and I need rectangle range query for simple visualization. the data is large enough to reject O(N) selection

Just a note of caution: Constructing a kd-tree will be $O(N\ log\  N)$ which is worse than $O(N)$.

**Comment by sturlamolden** [comment]
> I agree with Ralf that rectangular regions are fair game for SciPy 

Range query for hyperrectangles is very easy to implement, if the rectangle align with the axes. If they it does not it is a bit more tricky, without rebuilding the whole kd-tree (which defies the purpose, then a brute-force query is better).


> (but I think we can consider arbitrary polygons out of scope).

I have code for arbitrary 2D-polygons, if it is of interest (or maybe SciPy has that, I don't remember).

**Comment by mdhaber** [comment]
I am just asking for N-d hyperrectangles aligned with the axes. Thanks for considering it!
